h2o-core/src/main/java/water/JavaSelfCheck.java

```diff
@@ -0,0 +1,56 @@
+package water;
+
+import water.util.UnsafeUtils;
+
+import java.util.Random;
+
+public class JavaSelfCheck {
+
+    public static boolean checkCompatibility() {
+        if (!checkUnsafe())
+            return false;
+        if (!checkWeaver())
+            return false;
+        return checkSerialization();
+    }
+
+    private static boolean checkSerialization() {
+        byte[] bytes = AutoBuffer.serializeBootstrapFreezable(new HeartBeat());
+        return AutoBuffer.deserializeBootstrapFreezable(bytes) instanceof HeartBeat;
+    }
+
+    private static boolean checkWeaver() {
+        // HeartBeat class is guaranteed to be part of boostrap classes, otherwise clouding couldn't work
+        Freezable<?> f = TypeMap.newFreezable(HeartBeat.class.getName());
+        if (! (f instanceof HeartBeat)) {
+            return false;
+        }
+        // check that we can generate Icer
+        return TypeMap.getIcer(f) != null;
+    }
+
+    private static boolean checkUnsafe() {
+        final int N = 1024;
+        final Random r = new Random();
+        final double[] doubleData = new double[N];
+        final byte[] doubleBytes = new byte[N * 8];
+        final long[] longData = new long[N];
+        final byte[] longBytes = new byte[N * 8];
+        for (int i = 0; i < N; i++) {
+            doubleData[i] = r.nextDouble();
+            UnsafeUtils.set8d(doubleBytes, i * 8, doubleData[i]);
```

Comment:
Can this fail on something like UnsupportedOperationException from Unsafe itself?

Reply:
yes, it most likely can - this is accounted for - if we get any exception -> we are not compatible with the Java